File: generated_code/GeminiPro/Java/031_SplayTree.java
/**
 * A self-contained Java class that implements an integer set using a Splay Tree.
 * A splay tree is a self-balancing binary search tree with the additional property
 * that recently accessed elements are quick to access again. It performs basic
 * operations such as insertion, removal, and search in O(log n) amortized time.
 */
public class SplayTree {

    /**
     * Represents a node in the Splay Tree.
     */
    private class Node {
        int key;
        Node left, right;

        /**
         * Constructs a new node with the given key.
         * @param key The integer key for the node.
         */
        Node(int key) {
            this.key = key;
            this.left = null;
            this.right = null;
        }
    }

    private Node root;

    /**
     * Constructs an empty Splay Tree.
     */
    public SplayTree() {
        this.root = null;
    }

    /**
     * Performs a right rotation on the subtree rooted at node x.
     * @param x The root of the subtree to rotate.
     * @return The new root of the rotated subtree.
     */
    private Node rightRotate(Node x) {
        Node y = x.left;
        x.left = y.right;
        y.right = x;
        return y;
    }

    /**
     * Performs a left rotation on the subtree rooted at node x.
     * @param x The root of the subtree to rotate.
     * @return The new root of the rotated subtree.
     */
    private Node leftRotate(Node x) {
        Node y = x.right;
        x.right = y.left;
        y.left = x;
        return y;
    }

    /**
     * The splay operation. It brings the node with the given key (or the last
     * accessed node if the key is not found) to the root of the tree.
     * @param node The root of the tree/subtree to splay.
     * @param key The key to splay for.
     * @return The new root of the splayed tree.
     */
    private Node splay(Node node, int key) {
        if (node == null || node.key == key) {
            return node;
        }

        if (key < node.key) {
            // Key is in the left subtree
            if (node.left == null) {
                return node; // Key not found, splay the last accessed node
            }
            if (key < node.left.key) { // Zig-Zig (Left-Left)
                // Recursively bring the key as root of left-left
                node.left.left = splay(node.left.left, key);
                // First rotation for grandparent-parent
                node = rightRotate(node);
            } else if (key > node.left.key) { // Zig-Zag (Left-Right)
                // Recursively bring the key as root of left-right
                node.left.right = splay(node.left.right, key);
                // First rotation for parent-child if necessary
                if (node.left.right != null) {
                    node.left = leftRotate(node.left);
                }
            }
            // Second rotation for parent-child
            return (node.left == null) ? node : rightRotate(node);
        } else { // key > node.key
            // Key is in the right subtree
            if (node.right == null) {
                return node; // Key not found, splay the last accessed node
            }
            if (key > node.right.key) { // Zig-Zig (Right-Right)
                // Recursively bring the key as root of right-right
                node.right.right = splay(node.right.right, key);
                // First rotation for grandparent-parent
                node = leftRotate(node);
            } else if (key < node.right.key) { // Zig-Zag (Right-Left)
                // Recursively bring the key as root of right-left
                node.right.left = splay(node.right.left, key);
                // First rotation for parent-child if necessary
                if (node.right.left != null) {
                    node.right = rightRotate(node.right);
                }
            }
            // Second rotation for parent-child
            return (node.right == null) ? node : leftRotate(node);
        }
    }

    /**
     * Searches for a key in the tree. If the key is found, it is splayed to the root.
     * @param key The key to search for.
     * @return The key if found, otherwise null.
     */
    public Integer search(int key) {
        root = splay(root, key);
        if (root != null && root.key == key) {
            return root.key;
        }
        return null;
    }

    /**
     * Inserts a key into the splay tree. If the key already exists, the tree remains unchanged.
     * @param key The key to insert.
     */
    public void insert(int key) {
        // If tree is empty, create a new root.
        if (root == null) {
            root = new Node(key);
            return;
        }

        // Splay the tree to bring the closest node to the root.
        root = splay(root, key);

        // If key is already present, do nothing.
        if (root.key == key) {
            return;
        }

        Node newNode = new Node(key);

        if (key < root.key) {
            newNode.right = root;
            newNode.left = root.left;
            root.left = null;
        } else { // key > root.key
            newNode.left = root;
            newNode.right = root.right;
            root.right = null;
        }
        root = newNode;
    }

    /**
     * Deletes a key from the splay tree.
     * @param key The key to delete.
     */
    public void delete(int key) {
        if (root == null) {
            return;
        }

        // Splay the tree to bring the node to be deleted to the root.
        root = splay(root, key);

        // If the key is not in the tree, do nothing.
        if (root.key != key) {
            return;
        }

        Node leftSubtree = root.left;
        Node rightSubtree = root.right;

        if (leftSubtree == null) {
            root = rightSubtree;
        } else {
            // Splay the left subtree to bring the maximum element to its root.
            // This maximum element is the predecessor of the deleted key.
            // Searching for the key (which is guaranteed to be larger than any
            // element in the left subtree) will splay the max element to the top.
            leftSubtree = splay(leftSubtree, key);
            
            // Attach the right subtree to the new root (the max of the left subtree).
            leftSubtree.right = rightSubtree;
            root = leftSubtree;
        }
    }
}

File: generated_code/GeminiPro/Python/031_SplayTree.py
import sys

# It's good practice to increase recursion limit for deep trees, though splaying keeps the tree balanced.
sys.setrecursionlimit(2000)

class SplayTree:
    """
    An implementation of a Splay Tree.
    
    A splay tree is a self-adjusting binary search tree with the additional
    property that recently accessed elements are quick to access again.
    It performs basic operations such as insertion, look-up and removal
    in O(log n) amortized time.
    
    This class implements a dictionary-like set for storing unique integer keys.
    """

    class _Node:
        """A node in the Splay Tree."""
        def __init__(self, key, parent=None):
            self.key = key
            self.parent = parent
            self.left = None
            self.right = None

    def __init__(self):
        """Initializes an empty Splay Tree."""
        self.root = None

    def _left_rotate(self, x):
        """Performs a left rotation on node x."""
        y = x.right
        if y:
            x.right = y.left
            if y.left:
                y.left.parent = x
            y.parent = x.parent
        if not x.parent:
            self.root = y
        elif x == x.parent.left:
            x.parent.left = y
        else:
            x.parent.right = y
        if y:
            y.left = x
        x.parent = y

    def _right_rotate(self, x):
        """Performs a right rotation on node x."""
        y = x.left
        if y:
            x.left = y.right
            if y.right:
                y.right.parent = x
            y.parent = x.parent
        if not x.parent:
            self.root = y
        elif x == x.parent.right:
            x.parent.right = y
        else:
            x.parent.left = y
        if y:
            y.right = x
        x.parent = y

    def _splay(self, x):
        """
        Performs the splay operation on node x, moving it to the root.
        """
        while x.parent:
            p = x.parent
            g = p.parent
            if not g:  # Zig case
                if x == p.left:
                    self._right_rotate(p)
                else:
                    self._left_rotate(p)
            elif x == p.left and p == g.left:  # Zig-Zig case
                self._right_rotate(g)
                self._right_rotate(p)
            elif x == p.right and p == g.right:  # Zig-Zig case
                self._left_rotate(g)
                self._left_rotate(p)
            elif x == p.right and p == g.left:  # Zig-Zag case
                self._left_rotate(p)
                self._right_rotate(g)
            else:  # Zig-Zag case (x == p.left and p == g.right)
                self._right_rotate(p)
                self._left_rotate(g)

    def search(self, key):
        """
        Searches for a key in the tree.
        
        Performs the splaying operation on the accessed node if found.
        If the node is not found, its would-be parent is splayed to the root.
        
        Args:
            key: The integer key to search for.
            
        Returns:
            True if the key is found, False otherwise.
        """
        node = self.root
        last_node = None
        while node:
            last_node = node
            if key < node.key:
                node = node.left
            elif key > node.key:
                node = node.right
            else:  # Key found
                self._splay(node)
                return True
        
        # Key not found, splay the last visited node (parent of the would-be key)
        if last_node:
            self._splay(last_node)
        return False

    def insert(self, key):
        """
        Inserts a new key into the tree.
        
        If the key already exists, the existing node is splayed to the root.
        If the key is new, it is inserted and then splayed to the root.
        
        Args:
            key: The integer key to insert.
        """
        node = self.root
        parent = None
        while node:
            parent = node
            if key < node.key:
                node = node.left
            elif key > node.key:
                node = node.right
            else:  # Key already exists
                self._splay(node)
                return

        new_node = self._Node(key, parent)
        if not parent:
            self.root = new_node
        elif key < parent.key:
            parent.left = new_node
        else:
            parent.right = new_node
        
        self._splay(new_node)

    def delete(self, key):
        """
        Deletes a key from the tree.
        
        The tree is first searched for the key, which splays the found node
        (or its parent if not found) to the root. If the key is found,
        the node is removed and the tree is restructured.
        
        Args:
            key: The integer key to delete.
        """
        # Search for the key. This splays the node or its parent to the root.
        if not self.search(key):
            # Key not in tree. search() already splayed the parent, so we are done.
            return
        
        # At this point, the node to delete is the root because search() found it.
        z = self.root
        left_subtree = z.left
        right_subtree = z.right

        if not left_subtree:
            # Promote the right subtree to be the new tree
            self.root = right_subtree
            if right_subtree:
                right_subtree.parent = None
        else:
            # Detach the left subtree to operate on it
            left_subtree.parent = None
            
            # Find the maximum node in the left subtree
            max_node = left_subtree
            while max_node.right:
                max_node = max_node.right
            
            # Splay this max_node. It becomes the root of the tree formed by
            # the left subtree. We do this by temporarily setting the root.
            self.root = left_subtree
            self._splay(max_node)
            
            # Now, self.root is max_node. Re-attach the original right subtree.
            # max_node is guaranteed to have no right child at this point.
            self.root.right = right_subtree
            if right_subtree:
                right_subtree.parent = self.root

    def _inorder_traversal(self, node, nodes_list):
        """Helper for in-order traversal."""
        if node:
            self._inorder_traversal(node.left, nodes_list)
            nodes_list.append(node.key)
            self._inorder_traversal(node.right, nodes_list)

    def __str__(self):
        """Returns an in-order string representation of the tree's keys."""
        if not self.root:
            return "SplayTree()"
        nodes = []
        self._inorder_traversal(self.root, nodes)
        return f"SplayTree({', '.join(map(str, nodes))})"

    def __repr__(self):
        """Returns the string representation of the tree."""
        return str(self)

if __name__ == '__main__':
    # Example Usage
    st = SplayTree()

    # Insertions
    keys_to_insert = [10, 20, 30, 40, 50, 25]
    print(f"Inserting keys: {keys_to_insert}")
    for key in keys_to_insert:
        st.insert(key)
        print(f"After inserting {key}: root is {st.root.key}, tree is {st}")

    print("\n" + "="*30 + "\n")

    # Searches
    keys_to_search = [30, 15, 50]
    print(f"Searching for keys: {keys_to_search}")
    for key in keys_to_search:
        found = st.search(key)
        status = "found" if found else "not found"
        print(f"Searched for {key} ({status}): root is now {st.root.key}, tree is {st}")

    print("\n" + "="*30 + "\n")
    
    # Deletions
    keys_to_delete = [10, 40, 60]
    print(f"Deleting keys: {keys_to_delete}")
    for key in keys_to_delete:
        print(f"Attempting to delete {key}...")
        st.delete(key)
        if st.root:
            print(f"After deleting {key}: root is {st.root.key}, tree is {st}")
        else:
            print(f"After deleting {key}: tree is empty.")